rtsp.go

```diff
@@ -675,26 +669,33 @@ func newRTSPCamera(ctx context.Context, _ resource.Dependencies, conf resource.C
 	}
 	cancelCtx, cancel := context.WithCancel(context.Background())
 	reader := gostream.VideoReaderFunc(func(_ context.Context) (image.Image, func(), error) {
-		latest := rc.latestOutput.Load()
-		if latest == nil {
-			return nil, func() {}, errors.New("no frame yet")
+		if videoCodec(rc.currentCodec.Load()) == MJPEG {
+			img := rc.latestMJPEGImage.Load()
+			if img == nil {
+				return nil, func() {}, errors.New("no mjpeg frame yet")
+			}
+			return *img, func() {}, nil
 		}
 
-		release := func() {}
-		// When model is mjpeg, latest frame wrapper will be nil.
-		if latest.frameWrapper != nil {
-			latest.frameWrapper.isBeingServed.Store(true)
+		frame := rc.latestFrame.Load()
+		if frame == nil {
+			return nil, func() {}, errors.New("no frame yet")
+		}
 
-			release = func() {
-				// When the caller is done with the image, we return the AVFrame to the pool such
-				// that we can re-use its allocated byte buffer.
-				rc.logger.Debug("Release was called.")
-				latest.frameWrapper.isBeingServed.Store(false)
-				rc.avFramePool.safelyPut(latest.frameWrapper)
-			}
+		release := func() {
+			// When the caller is done with the image, we return the AVFrame to the pool such
+			// that we can re-use its allocated byte buffer.
+			rc.logger.Debug("Release was called.")
+			frame.isBeingServed.Store(false)
+			frame.refs.Add(-1)
```

Comment:
`put` does the `Add(-1)` -- I think this is a double deref bug?

Reply:
My understanding is that this `Add(-1)`'s equivalent `Add(1)` is inside of this `VideoReaderFunc` defer.

The one in `put`'s equivalent is in `get` (or when we make a new frame when the `srcYUV420Frame`/`dst` dimensions don't match).

Reply:
My understanding was incorrect. Removed extra deref inside of `put`. In fact, there is now no ref handling at all in pool.go